### ISSUE TITLE
关于the matte loss损失及数据集image的normalized问题

### ISSUE THREAD

**Issue by chenying99** [ZHKKKe/MODNet#133]
感谢作者分享！请教以下问题：
 
   1） 关于matte loss损失  (trainer.py文件的160行)

    # calculate the matte loss
    pred_boundary_matte = torch.where(boundaries, trimap, pred_matte)
    matte_l1_loss = F.l1_loss(pred_matte, gt_matte) + 4.0 * F.l1_loss(pred_boundary_matte, gt_matte)
    
pred_boundary_matte与gt_matte的**非boundaries**区域的值应该是一致的吧？
（注：# calculate the boundary mask from the trimap
           boundaries = (trimap < 0.5) + (trimap > 0.5)） 感觉语义上应该取反，貌似更合理）
后半部分的损失相当于只针对该区域的损失的进行计算？ 整体上（两部分）相当于对该区域的损失进行了加权计算？我的理解是否正确？

而且后面的均值，matte_loss = torch.mean(matte_l1_loss + matte_compositional_loss)：
如果4.0 * F.l1_loss(pred_boundary_matte, gt_matte)只计算特定区域，这里的均值是针对全体的，包括其他区域（其大小每个样本不一致），这回导致均值不稳定吧？

matte_compositional_loss = F.l1_loss(image * pred_matte, image * gt_matte) \
        + 4.0 * F.l1_loss(image * pred_boundary_matte, image * gt_matte)
这部分损失计算（matte_compositional_loss）是不是有点冗余呢？相当于对上一个损失计算（matte_l1_loss）进行了scale，乘以image因子？  是不是可以合并到上一个损失计算（matte_l1_loss）里面，适当加权？ 

2）请教训练时image是normalized到[0, 1]还是[-1, 1]呢？ 是不是两者都可以？谢谢！

**Comment by ZHKKKe** [comment]
你好，感谢你的关注。
对于你的问题：

**Q1: 后半部分的损失相当于只针对该区域的损失的进行计算？ 整体上（两部分）相当于对该区域的损失进行了加权计算？我的理解是否正确?**
是的，你的理解正确。

**Q2: 如果4.0 * F.l1_loss(pred_boundary_matte, gt_matte)只计算特定区域，这里的均值是针对全体的，包括其他区域（其大小每个样本不一致），这回导致均值不稳定吧?**
抱歉，由于我们对边缘区域的损失增加权重的过程中没有遇到均值不稳定的问题，可能无法提供有效的建议。

**Q3: 这部分损失计算（matte_compositional_loss）是不是有点冗余呢？相当于对上一个损失计算（matte_l1_loss）进行了scale，乘以image因子？ 是不是可以合并到上一个损失计算（matte_l1_loss）里面，适当加权？**
matte_compositional_loss 并不是一个主要的损失，去除它对最终结果的影响可能较小。我们在工作中follow之前的工作加上了这一损失。

**Q4: 请教训练时image是normalized到[0, 1]还是[-1, 1]呢？ 是不是两者都可以？**
在我们的实验里，输入image归一化到[0, 1]还是[-1, 1]对结果影响不大。
